server/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/AvgAggregator.java

```diff
@@ -76,15 +78,29 @@ public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
             public void collect(int doc, long bucket) throws IOException {
                 counts = bigArrays.grow(counts, bucket + 1);
                 sums = bigArrays.grow(sums, bucket + 1);
+                compensations = bigArrays.grow(compensations, bucket + 1);
 
                 if (values.advanceExact(doc)) {
                     final int valueCount = values.docValueCount();
                     counts.increment(bucket, valueCount);
-                    double sum = 0;
+                    // Compute the sum of double values with Kahan summation algorithm which is more
+                    // accurate than naive summation.
+                    double sum = sums.get(bucket);
+                    double compensation = compensations.get(bucket);
+
                     for (int i = 0; i < valueCount; i++) {
-                        sum += values.nextValue();
+                        double value = values.nextValue();
+                        if (Double.isFinite(value) == false) {
+                            sum += value;
+                        } else if (Double.isFinite(sum)) {
+                            double corrected = value - compensation;
+                            double newSum = sum + corrected;
+                            compensation = (newSum - sum) - corrected;
+                            sum = newSum;
+                        }
```

Comment:
I think one corner-case is not covered with this logic. Imagine that all values are finite but 2: -Inf and +Inf, for instance:
`[+Inf, 4, -Inf]`. The expected result is `NaN` but your logic will make it return `+Inf` if I'm not mistaken. Maybe this should be something like that instead:

```java
if (Double.isFinite(value) == false || Double.isFinite(sum) == false) {
  sum += value;
} else {
  // kahan summation
}
```

Reply:
Hi @jpountz Thanks a lot! I made a test for both of them:
```
        double sum = 0;
        double compensation = 0;

        double[] values = new double[]{Double.POSITIVE_INFINITY, 4, Double.NEGATIVE_INFINITY};

        for (int i = 0; i < values.length; i++) {
            double value = values[i];
            if (Double.isFinite(value) == false) {
                sum += value;
            } else if (Double.isFinite(sum)) {
                double corrected = value - compensation;
                double newSum = sum + corrected;
                compensation = (newSum - sum) - corrected;
                sum = newSum;
            }
        }
        System.out.println(sum);

        sum = 0;
        compensation = 0;

        for (int i = 0; i < values.length; i++) {
            double value = values[i];
            if (Double.isFinite(value) == false || Double.isFinite(sum) == false) {
                sum += value;
            } else {
                double corrected = value - compensation;
                double newSum = sum + corrected;
                compensation = (newSum - sum) - corrected;
                sum = newSum;
            }
        }
        System.out.println(sum);
```
Both of their result are `NaN`. Because in my code, for each value, if it's not finite, it will be summed up to `sum`, no matter what `sum` is. I can also made this change because I also think your way is more intuitive. I also have a random test case for this kind corner-case:
```
        int n = randomIntBetween(5, 10);
        values = new double[n];
        double sum = 0;
        for (int i = 0; i < n; i++) {
            values[i] = frequently()
                ? randomFrom(Double.NaN, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY)
                : randomDoubleBetween(Double.MIN_VALUE, Double.MAX_VALUE, true);
            sum += values[i];
        }
        verifyAvgOfDoubles(values, sum / n, 1e-10);
```
I can make it be regular for example:
```
        double[] values = new double[]{Double.POSITIVE_INFINITY, 4, Double.NEGATIVE_INFINITY};
        verifyAvgOfDoubles(values, NaN, 0d);
```
WDYT?

Reply:
I had misread your code, sorry! I am fine either way.

Reply:
Let's keep things the way they are.